ci/ray_ci/doc/api.py

```diff
@@ -118,3 +118,20 @@ def get_canonical_name(self) -> str:
             attribute = getattr(attribute, token)
 
         return f"{attribute.__module__}.{attribute.__qualname__}"
+
+    def _is_private_name(self) -> bool:
+        """
+        Check if this API has a private name. Private names are those that start with
+        underscores.
+        """
+        return self.name.split(".")[-1].startswith("_")
```

Comment:
there are many special functions in python that are actually public, but starts with `_`, like `__del__`, `__iter__`, `__len__`, `__init__`, `__next__`, etc, etc..

Reply:
and I am actually not sure about the annotation story for those in ray's api world..

Reply:
> not sure about the annotation story for those in ray's api worl

that is fine, and why we need to agree on this api policy across teams (see line 8 which is our agreed definition of private, not public, etc.)

<img width="1139" alt="Screenshot 2024-06-25 at 9 24 00 PM" src="https://github.com/ray-project/ray/assets/128072568/86268a58-e01b-4fc9-b1b2-554df3c1ef0c">

i will be implementing what is in this policy for now, and if the policy change i will update the code accordingly; at least this will allow me to turn on the lint check for data team who has aligned with this policy

Reply:
what does data team say about these api's? maybe ask them to have a look?

like at least `__init__` method of a public api class needs to be public, right? do those also need to be explicitly annotated?

Reply:
currenly folks do not annotate the `__init__` methods, or any other `_` methods; instead the class needs to be annotated (e.g. https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L130)

existing logic for what is excluded from annotations https://github.com/ray-project/ray/blob/master/ci/lint/check_api_annotations.py#L42 (`_` names and things under `.internal.`)

Reply:
in documentation, the class is used instead of the `__init__` method as the constructor https://docs.ray.io/en/latest/data/api/dataset.html (e.g. `ray.data.Dataset` instead of `ray.data.Dataset.__init__`)

Reply:
> at least this will allow me to turn on the lint check for data team who has aligned with this policy

The table overall looks good to me. To clarify, we're allowed to completely deprecate a stable API after 6 months warning, but we're not allowed to deprecate a parameter of a stable API with 6 months warning?

> currenly folks do not annotate the __init__ methods, or any other _ methods; instead the class needs to be annotated 

Yeah, we don't typically annotate dunder method. It's conventional to document either `__init__` or the class itself, but I think we annotate the class more often (if not always).